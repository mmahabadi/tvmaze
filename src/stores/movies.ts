import { defineStore } from 'pinia'
import { computed, ref, type ComputedRef } from 'vue'
import * as ApiService from '@/services/ApiService'
import { type Movie, type ShowByGenre } from '@/types'

export const useMoviesStore = defineStore('movies', () => {
  const shows = ref<Map<string, ShowByGenre>>(new Map())
  const movies = ref<Movie[]>([])
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)
  const currentIndex = ref(1)
  const ITEMS_PER_PAGE = 3
  const ITEMS_LIMIT_FOR_RECENT_MOVIES = 15
  const DATE_LIMIT_FOR_RECENT_MOVIES = 2020

  const recentMovies = computed(() =>
    movies.value
      .filter((movie) => new Date(movie.ended).getFullYear() > DATE_LIMIT_FOR_RECENT_MOVIES)
      .slice(0, ITEMS_LIMIT_FOR_RECENT_MOVIES)
  )

  /**
   * @note because the API does not support fetching genres,
   * I need to extract the genres from the fetched movies
   */
  const genres = computed(() => [...new Set([...movies.value.flatMap((movie) => movie.genres)])])

  const getByGenre = computed(() => (genre: string): ShowByGenre => {
    loadGenre(genre)
    return shows.value.get(genre) as ShowByGenre
  })

  const paginatedGenres: ComputedRef<string[]> = computed(() => {
    const startIndex = currentIndex.value
    const endIndex = startIndex * ITEMS_PER_PAGE
    // Concatenate the new slice with the previous values of pagedGenres
    return [...genres.value.slice(0, endIndex)]
  })

  async function fetchMovies(): Promise<void> {
    loading.value = true
    try {
      movies.value = await ApiService.getData()
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  async function fetchShowByGenre(genre: string): Promise<void> {
    const genreShows = shows.value.get(genre)
    if (!genreShows) return

    try {
      /**
       * @note if the APi supports fetching movies by genre,
       * comment the following line
       * and uncomment the line below
       */
      const items = movies.value
        .filter((movie) => movie.genres.includes(genre))
        .sort((a, b) => b.rating.average - a.rating.average)
      /**
       * @note the API does not support fetching movies by genre
       * uncomment the following line to fetch the movies by genre from the API
       */
      // const items = await ApiService.getShowsByGenre(genre)
      if (items) {
        genreShows.movies = items
      }
    } catch (e) {
      genreShows.error = e as Error
    } finally {
      genreShows.loading = false
    }
    shows.value.set(genre, genreShows as ShowByGenre)
  }

  function loadGenre(genre: string): void {
    if (!shows.value.has(genre)) {
      const genreShows: ShowByGenre = {
        genre,
        movies: [],
        loading: true,
        error: null
      }
      shows.value.set(genre, genreShows)
      fetchShowByGenre(genre)
    }
  }

  function nextPage(): void {
    if (currentIndex.value * ITEMS_PER_PAGE <= genres.value.length) currentIndex.value++
  }

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    movies,
    genres: paginatedGenres,
    recentMovies,
    details,
    loading,
    error,
    fetchMovies,
    getByGenre,
    nextPage,
    getShowDetails
  }
})
