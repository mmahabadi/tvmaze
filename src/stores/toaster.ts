import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { ToasterMessage } from '@/types'

export const useToasterStore = defineStore('toaster', () => {
  const toasters = ref<ToasterMessage[]>([])
  let idCounter = 0

  const addToaster = (message: string, type: 'warning' | 'error' | 'success') => {
    toasters.value.push({ id: idCounter++, message, type })
  }

  const removeToaster = (id: number) => {
    toasters.value = toasters.value.filter((toaster) => toaster.id !== id)
  }

  return {
    toasters,
    addToaster,
    removeToaster
  }
})
