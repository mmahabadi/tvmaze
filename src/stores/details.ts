import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'
import type { Movie } from '@/types'
import { useToasterStore } from './toaster'

export const useDetailsStore = defineStore('details', () => {
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)
  const { addToaster } = useToasterStore()

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
      addToaster(`Failed to fetch the show`, 'error')
    } finally {
      loading.value = false
    }
  }

  return {
    details,
    loading,
    error,
    getShowDetails
  }
})
