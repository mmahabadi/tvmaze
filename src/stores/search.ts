import type { Movie } from '@/types'
import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'
import { useToasterStore } from './toaster'

export const useSearchStore = defineStore('search', () => {
  const searchResults = ref<Movie[]>([])
  let searchQuery = ''
  const loading = ref(false)
  const error = ref<null | Error>(null)

  const { addToaster } = useToasterStore()

  async function searchShows(query: string): Promise<void> {
    //avoid making the same request multiple times
    if (searchQuery === query) return
    searchQuery = query
    loading.value = true
    try {
      searchResults.value = await ApiService.search(query)
    } catch (e) {
      error.value = e as Error
      addToaster(`Failed to fetch shows`, 'error')
    } finally {
      loading.value = false
    }
  }

  return {
    searchResults,
    searchQuery,
    loading,
    error,
    searchShows
  }
})
