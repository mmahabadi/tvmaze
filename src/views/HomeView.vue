<script setup lang="ts">
import { onMounted } from 'vue'
import { useMoviesStore } from '@/stores/movies'
import MainLayout from '@/layouts/MainLayout.vue'
import ImageSlider from '@/components/slider/ImageSlider.vue'
import TopRatedMovies from '@/components/TopRatedMovies.vue'
import RecentMovies from '@/components/RecentMovies.vue'
import CTA from '@/components/CTA.vue'

const store = useMoviesStore()

//TODO: add error handling
//TODO: add skeleton loading
//TODO: add lazy
//TODO: add proxy

onMounted(() => {
  store.fetchMovies()
})
</script>

<template>
  <MainLayout>
    <template v-slot:slider>
      <ImageSlider />
    </template>
    <TopRatedMovies />
    <CTA />
    <RecentMovies />
  </MainLayout>
</template>
