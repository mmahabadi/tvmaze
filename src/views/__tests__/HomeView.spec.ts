import { setActivePinia, createPinia } from 'pinia'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import HomeView from '../HomeView.vue'
import { useMoviesStore } from '@/stores/movies'
import type { Movie } from '@/types'

const mockMovies: Movie[] = [
  {
    id: 1,
    name: 'Inception',
    genres: ['actions', 'comedy'],
    image: {
      medium: 'inception.jpg',
      original: 'inception.jpg'
    }
  } as Movie
]
describe('HomeView', () => {
  let wrapper: VueWrapper<HomeView>
  let store: ReturnType<typeof useMoviesStore>

  function render(store: ReturnType<typeof useMoviesStore>) {
    if (wrapper) wrapper.unmount()

    wrapper = mount(HomeView, {
      global: {
        stubs: {
          ImageSlider: true,
          ThumbnailSwiper: true,
          MainLayout: true,
          ListByGenre: true
        },
        provide: {
          store
        }
      }
    })
  }

  beforeEach(() => {
    setActivePinia(createPinia())

    store = useMoviesStore()
    store.movies = mockMovies
    store.fetchMovies = vi.fn()

    render(store)
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('calls fetchMovies on mount if genres are empty', () => {
    //Arrange
    store.movies = []
    const fetchMoviesSpy = vi.spyOn(store, 'fetchMovies')

    //Act
    render(store)

    //Assert
    expect(fetchMoviesSpy).toHaveBeenCalled()
  })

  it('does not call fetchMovies on mount if genres are not empty', () => {
    //Arrange
    store.movies = mockMovies
    const fetchMoviesSpy = vi.spyOn(store, 'fetchMovies')

    //Act
    render(store)

    expect(fetchMoviesSpy).not.toHaveBeenCalled()
  })

  it('calls nextPage when scrolled to the bottom', () => {
    const nextPageSpy = vi.spyOn(store, 'nextPage')

    // Mock scroll event
    document.documentElement.scrollTop = 1000
    Object.defineProperty(document.documentElement, 'clientHeight', { value: 600 })
    Object.defineProperty(document.documentElement, 'scrollHeight', { value: 1600 })

    const event = new Event('scroll')
    document.dispatchEvent(event)

    expect(nextPageSpy).toHaveBeenCalled()
  })
})
