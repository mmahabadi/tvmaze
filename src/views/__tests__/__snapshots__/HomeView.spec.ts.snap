// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeView > renders correctly 1`] = `"<main-layout-stub loading="false"></main-layout-stub>"`;
